### PR TITLE
CDAP-3854 Adding changes to configure kafka topic to which logs and n…

### PR DESCRIPTION
**Pull request by CuriousVini**
…otifications are published

JIRA - https://issues.cask.co/browse/CDAP-3854
Build - http://builds.cask.co/browse/CDAP-DUT3561-9

### CODE REVIEW

File: cdap-common/src/main/resources/cdap-default.xml [hunk 1]
```diff
@@ -835,6 +835,14 @@
...
   </property>
 
+  <property>
+    <name>log.kafka.topic.prefix</name>
+    <value>logs.user-v2</value>
+    <description>
+      Kafka log topic to publish logs
+    </description>
+  </property>
+
```
Comment (awholegunch) [inline cdap-common/src/main/resources/cdap-default.xml]: These properties are inserted in alphabetical order in each section. So, this one should be earlier.

Is that default value correct? Just wondering...I would suggest for the description:
```
Kafka topic prefix used to publish logs

File: cdap-common/src/main/resources/cdap-default.xml [hunk 2]
```diff
@@ -1200,6 +1208,14 @@
...
   </property>
 
+  <property>
+    <name>notification.kafka.topic.prefix</name>
+    <value>notifications</value>
+    <description>
+      Kafka topic name to publish notifications
+    </description>
+  </property>
+
```
Comment (awholegunch) [inline cdap-common/src/main/resources/cdap-default.xml]: These properties are inserted in alphabetical order in each section. So, this one should be before ``notification.transport.system``.

I would suggest for the description:
```
Kafka topic prefix used to publish notifications

File: cdap-common/src/main/resources/cdap-default.xml (hunk 1 above)
Comment (awholegunch) [inline cdap-common/src/main/resources/cdap-default.xml]: Also, why are these "KAFKA_TOPIC_PREFIX", when in the code, they are always assigned to a Kafka Topic? Should property just drop the ".prefix" portion? The descriptions would then be "Kafka topic to publish logs"...

**Comment by awholegunch** [comment]
Files with existing copyrights that are modified need to go from "2014" to "2014-2016" (common).

File: cdap-common/src/main/resources/cdap-default.xml (hunk 1 above)
Comment (awholegunch) [inline cdap-common/src/main/resources/cdap-default.xml]: Is this being used as a version string? ("v2") Perhaps add something to the property name or description so that it is known.

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java [hunk 3]
```diff
@@ -79,8 +81,9 @@ public KafkaNotificationService(CConfiguration cConf, KafkaClient kafkaClient, D
     this.feedManager = feedManager;
     this.ack = KafkaPublisher.Ack.LEADER_RECEIVED;
     this.nbPartitions = cConf.getInt("kafka.num.partitions");
-
+    this.notificationTopic = cConf.get(Constants.Notification.KAFKA_TOPIC_PREFIX);
```
Comment (bdmogal) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java]: final

File: cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java [hunk 4]
```diff
@@ -370,7 +370,7 @@ public static StandaloneMain create(CConfiguration cConf, Configuration hConf) {
       new DataSetsModules().getStandaloneModules(),
       new DataSetServiceModules().getStandaloneModules(),
       new MetricsClientRuntimeModule().getStandaloneModules(),
-      new LoggingModules().getStandaloneModules(),
+      new LoggingModules(configuration).getStandaloneModules(),
```
Comment (bdmogal) [inline cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java]: would renaming this variable to ``cConf`` involve too many changes?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java [hunk 5]
```diff
@@ -42,11 +43,11 @@
    * Creates a DistributedLogReader object.
    */
   @Inject
-  public DistributedLogReader(KafkaLogReader kafkaLogReader, FileLogReader fileLogReader,
+  public DistributedLogReader(@Named(Constants.Logging.KAFKA_TOPIC_PREFIX) String kafkaTopic, KafkaLogReader kafkaLogReader, FileLogReader fileLogReader,
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java]: Checkstyle error.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/KafkaLogReader.java [hunk 6]
```diff
@@ -55,13 +56,13 @@
    * @param cConfig configuration object containing Kafka seed brokers and number of Kafka partitions for log topic.
    */
   @Inject
-  public KafkaLogReader(CConfiguration cConfig, StringPartitioner partitioner) {
+  public KafkaLogReader(@Named(Constants.Logging.KAFKA_TOPIC_PREFIX) String kafkaTopic, CConfiguration cConfig, StringPartitioner partitioner) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/KafkaLogReader.java]: Same here. Checkstyle error. Line too long.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java (hunk 5 above)
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java]: can this be made non-public?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java [hunk 7]
```diff
@@ -40,11 +42,11 @@
   private final AtomicBoolean stopped = new AtomicBoolean(false);
 
   @Inject
-  public KafkaLogAppender(CConfiguration configuration) {
+  public KafkaLogAppender(CConfiguration configuration, @Named(Constants.Logging.KAFKA_TOPIC_PREFIX) String kafkaTopic) {
```
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java]: can this be made non-public?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/KafkaLogReader.java (hunk 6 above)
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/KafkaLogReader.java]: ditto. Also, can we rename the param to ``cConf`` here too?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java [hunk 8]
```diff
@@ -45,6 +47,12 @@
  */
 public class LoggingModules extends RuntimeModule {
 
+  private final CConfiguration cConf;
+
+  public LoggingModules(CConfiguration cConf) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java]: Shouldn't take a `CConfiguration` in constructor, but it should be acquired through Guice Injector

1. It involves a lot of changes
2. There can be configs set/add through the ConfigModule. Pass it directly will lack those changes.

The change would be simpler if you just have the corresponding reader (e.g. DistributedLogReader, KafkaLogReader) to take a `CConfiguration` in the constructor as those constructors are already annotated with `@Inject`.

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/TestDistributedLogReader.java [hunk 9]
```diff
@@ -78,6 +78,7 @@ public static void setUpContext() throws Exception {
     cConf.setInt(LoggingConfiguration.LOG_MAX_FILE_SIZE_BYTES, 20 * 1024);
     cConf.set(LoggingConfiguration.NUM_PARTITIONS, "2");
     cConf.set(LoggingConfiguration.KAFKA_PRODUCER_TYPE, "sync");
+    cConf.set(Constants.Logging.KAFKA_TOPIC_PREFIX, "logs.user-v2");
```
Comment (bdmogal) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/TestDistributedLogReader.java]: since its part of ``cdap-default.xml``, do we have to set this explicitly?

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java (hunk 3 above)
Comment (anew) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java]: Do Kafka topics allow arbitrary strings? Or do we need to validate that it is a valid topic name?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java [hunk 10]
```diff
@@ -40,11 +42,11 @@
...
     setName(APPENDER_NAME);
     addInfo("Initializing KafkaLogAppender...");
 
-    this.producer = new SimpleKafkaProducer(configuration);
+    this.producer = new SimpleKafkaProducer(configuration, kafkaTopic);
```
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java]: so can we read the kafka topic directly from the provided ``cConf`` object? Instead of injecting it in the constructor? Or did I miss something?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java (hunk 7 above)
Comment (anew) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java]: why is this needed? isn't it in the CConf, too?

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/KafkaTestBase.java [hunk 11]
```diff
@@ -31,6 +32,8 @@
 import com.google.inject.assistedinject.FactoryModuleBuilder;
 import org.junit.ClassRule;
 
+import java.io.File;
```
Comment (anew) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/KafkaTestBase.java]: what is File needed for?

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java (hunk 3 above)
Comment (CuriousVini) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java]: Thanks @anew for the comments. Kafka topics allow ASCII alphanumerics, '.', '_' and '-', as shown [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/common/Topic.scala). If we validate Kafka topics, then our string validation will depend on Kafka's implementation. Instead, can we depend on Kafka to validate strings? Please let me know your thoughts.

Comment (anew) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java]: Does Kafka have a public method validate the string? If so, I would definitely recommend to use that.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java [hunk 12]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2014-2015 Cask Data, Inc.
+ * Copyright © 2014-2016 Cask Data, Inc.
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java]: If there is no modification to the file, no need to modify the Copyright statement.

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java [hunk 13]
```diff
@@ -66,12 +67,13 @@
...
   private final int nbPartitions;
+  private final String notificationTopic;
 
   // Executor to publish notifications to Kafka
   private ListeningExecutorService publishingExecutor;
 
   @Inject
-  public KafkaNotificationService(CConfiguration cConf, KafkaClient kafkaClient, DatasetFramework dsFramework,
+  KafkaNotificationService(CConfiguration cConf, KafkaClient kafkaClient, DatasetFramework dsFramework,
                                   TransactionSystemClientService transactionSystemClient,
```
Comment (chtyim) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java]: Mis-alignment.

File: cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java [hunk 14]
```diff
@@ -125,4 +137,14 @@ private boolean isPositiveInteger(String key) {
       return false;
     }
   }
+
+  private boolean isValidKafkaTopic(String kafkaTopic) {
+    try {
+      Topic.validate(kafkaTopic);
+    } catch (Exception e) {
+      LOG.error("  {} is not a valid kafka topic", kafkaTopic);
```
Comment (rohitsinha54) [inline cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java]: Why the space in beginning ?

File: cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java [hunk 15]
```diff
@@ -118,6 +120,16 @@ private void checkPotentialPortConflicts() {
     }
   }
 
+  // checks that kafka topic provided is a valid kafka topic
```
Comment (rohitsinha54) [inline cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java]: We don't put function docs with //. Will be better if change it to /* for consistency or just get rid of it as its a private method and the method name is self descriptive.

File: cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java (hunk 14 above)
Comment (rohitsinha54) [inline cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java]: Ignore. As it seems that they are there to group the messages.

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationUtils.java [hunk 16]
```diff
@@ -30,12 +30,12 @@
    * @param feed {@link Id.NotificationFeed} object
```
Comment (rohitsinha54) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationUtils.java]: Param should be added here.

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationUtils.java [hunk 17]
```diff
@@ -30,12 +30,12 @@
...
    * @return Kafka topic that should contain the Notifications published on the {@code feed}
    */
-  public static TopicPartition getKafkaTopicPartition(Id.NotificationFeed feed) {
+  public static TopicPartition getKafkaTopicPartition(String topic, Id.NotificationFeed feed) {
     // For now, we only have a topic per feed Category.
     // Later, we may want to have multiple topics per categories, defined in cdap-site.
     // For example, we may have 10 topics for the category streams, which names would be
     // notifications-streams-1 .. notifications-streams-10.
```
Comment (rohitsinha54) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationUtils.java]: Does this comment needs updating ?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java [hunk 18]
```diff
@@ -40,11 +40,11 @@
   private final AtomicBoolean stopped = new AtomicBoolean(false);
 
   @Inject
-  public KafkaLogAppender(CConfiguration configuration) {
```
Comment (rohitsinha54) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java]: Not your PR but APPENDER_NAME can be private. Can you please change that.

**Comment by rohitsinha54** [comment]
Few nits but  besides that LGTM. Please squash minor commits before merge

File: cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Tests Configuration for Kafka topic
+ */
+public class ConfigurationCheckTest {
+  private static ConfigurationCheck configurationCheck;
+
+  @BeforeClass
+  public static void setContext() {
+    CConfiguration cConf = CConfiguration.create();
+    cConf.set(Constants.Logging.KAFKA_TOPIC, "invalid:topic");
```
Comment (bdmogal) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: nit: there is a small problem in this test. It tests for either one of these configurations to be valid, but not both, right? To make it stricter, we should validate with different cConf objects, one that has an invalid Log Kafka topic, another that has an invalid Notifications Kafka Topic.

File: cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.guice.ConfigModule;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * Tests Configuration for Kafka topic
```
Comment (bdmogal) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: nit: The class name suggests that there would be tests for all checks in ``ConfigurationCheck``, not just for Kafka topic params.

File: cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java [hunk 21]
```diff
@@ -118,6 +120,16 @@ private void checkPotentialPortConflicts() {
...
+  private void checkKafkaTopic(Set<String> problemKeys) {
+    if (!isValidKafkaTopic(cConf.get(Constants.Notification.KAFKA_TOPIC))) {
+      problemKeys.add(Constants.Notification.KAFKA_TOPIC);
+    }
+    if (!isValidKafkaTopic(cConf.get(Constants.Logging.KAFKA_TOPIC))) {
+      problemKeys.add(Constants.Logging.KAFKA_TOPIC);
+    }
```
Comment (bdmogal) [inline cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java]: we should also check the ``Constants.Metadata.UPDATES_KAFKA_TOPIC`` here.

File: cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java (hunk 20 above)
Comment (CuriousVini) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: For now, it only has tests for Kafka topic validations. Later, we can add tests for other configurations as well

Comment (bdmogal) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: Yes, that's fine. But please update the javadoc to something like ``Tests for {@link ConfigurationCheck}``

File: cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java (hunk 19 above)
Comment (CuriousVini) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: As ConfigurationCheck.java throws RuntimeException whenever `problemKeys` set is non-empty. It will throw exception if any of the 3(Logs, Notification or Metadata) kafka topics are invalid and the test will fail.

Comment (bdmogal) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: Correct, it will fail if any of these are invalid. So my point was - though you set both ``Constants.Logging.KAFKA_TOPIC`` and ``Constants.Notifications.KAFKA_TOPIC`` to invalid values, you're producing the same effect as setting only one of them in the test case, right? The test case does not verify that the size of ``problemKeys`` is 2, so its not strict enough. e.g., if the check for ``Constants.Logging.KAFKA_TOPIC`` gets removed by accident in future, this test won't fail, therefore won't catch the accident/mistake, right?

File: cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java (hunk 14 above)
Comment (bdmogal) [inline cdap-master/src/main/java/co/cask/cdap/master/startup/ConfigurationCheck.java]: nit: the (Topic.validate)[https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/common/Topic.scala#L30] method has descriptive error messages. We should log the messages too.

Also, perhaps only catch ``InvalidTopicException`` here?

File: cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Tests for {@link ConfigurationCheck}
+ * For now, it only tests Kafka topic validations. Later, tests for other configurations can be added.
+ */
+public class ConfigurationCheckTest {
+
+  @Test (expected = RuntimeException.class)
+  public void invalidLoggingKafkaTopicTest() {
+    CConfiguration cConf = CConfiguration.create();
```
Comment (bdmogal) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: instead of this repetitive code, can you create a private method ``runConfigurationCheck(String param, String invalidValue)`` and call it with different values from each of these tests?

File: cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.guice.ConfigModule;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import org.junit.Test;
+
+/**
+ * Tests for {@link ConfigurationCheck}
+ * For now, it only tests Kafka topic validations. Later, tests for other configurations can be added.
```
Comment (bdmogal) [inline cdap-master/src/test/java/co/cask/cdap/master/startup/ConfigurationCheckTest.java]: nit: remove this line. Otherwise someone will have to remove it when they add more tests.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/SimpleKafkaProducer.java [hunk 24]
```diff
@@ -36,25 +37,24 @@
...
     Properties props = new Properties();
-    props.setProperty("metadata.broker.list", configuration.get(LoggingConfiguration.KAFKA_SEED_BROKERS));
+    props.setProperty("metadata.broker.list", cConf.get(LoggingConfiguration.KAFKA_SEED_BROKERS));
     props.setProperty("serializer.class", "kafka.serializer.DefaultEncoder");
     props.setProperty("key.serializer.class", "kafka.serializer.StringEncoder");
     props.setProperty("partitioner.class", "co.cask.cdap.logging.appender.kafka.StringPartitioner");
     props.setProperty("request.required.acks", "1");
-    props.setProperty("producer.type", configuration.get(LoggingConfiguration.KAFKA_PRODUCER_TYPE,
+    props.setProperty("producer.type", cConf.get(LoggingConfiguration.KAFKA_PRODUCER_TYPE,
                                                          LoggingConfiguration.DEFAULT_KAFKA_PRODUCER_TYPE));
```
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/SimpleKafkaProducer.java]: nit: indentation

Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/SimpleKafkaProducer.java]: infact, can you fix indentation in the rest of this method too?

**Comment by bdmogal** [comment]
@CuriousVini final few comments.

**Comment by bdmogal** [comment]
LGTM, pending green build and testing in all modes. Please squash commits before merging.
